### PR TITLE
Improve the heartbeat processing by not remove region from the tree

### PR DESCRIPTION
**Pull request by mantuliu**
### What problem does this PR solve? 
UCP: #2431 
Improve the heartbeat processing

### What is changed and how it works?
When removing the region, try not to remove the region from the region tree,

The test conditions are: 1 million region 20store 0.1 region-update-ratio, 10heartbeat-rounds. The average heartbeat is the average of the last 9 times.

the improve percent like this:
no.|improve content |related code |average heartbeat/s | improve percent
---|---|---|---|---
1|   master branch |     |   101652 | 0%    
2|use ShallowClone when UpdateStoreStatus.|   store.go 666: newStore := store.ShallowClone(SetLeaderCount(leaderCount),  | 124109   |     22%
3|use reduce the number of updates when update related stores.|   cluster.go 572-583: c.updateStoreStatusLocked(key),  | 132076   |     7.8%
4|try not to remove the region from the subTree.|   region.go 580-584: shouldRemoveFromSubTree,  | 151024   |     18.6%
5|try not to remove the region from the tree.|   region.go 610-629: AddRegion,  | 168702   |     17.4%

Tests 

 - Unit test

### CODE REVIEW

File: server/core/region.go [hunk 1]
```diff
@@ -605,12 +607,27 @@ func (r *RegionsInfo) GetOverlaps(region *RegionInfo) []*RegionInfo {
 
 // AddRegion adds RegionInfo to regionTree and regionMap, also update leaders and followers by region peers
 func (r *RegionsInfo) AddRegion(region *RegionInfo) []*RegionInfo {
-	// Add to tree and regions.
-	overlaps := r.tree.update(region)
-	for _, item := range overlaps {
-		r.RemoveRegion(r.GetRegion(item.GetID()))
+	var overlaps []*RegionInfo
+	treeNeedAdd := true
```
Comment (nolouch) [inline server/core/region.go:611]: Could you add a comment about the variable.

Comment (mantuliu) [reply server/core/region.go:613]: Done, thanks.

**Pull request merged**